### PR TITLE
Cells to baronies

### PR DESCRIPTION
**Pull request by MnTronslien**
Reworking the title generation to better match what I think is the ideal way to convert the data.
Burgs > baronies
Counties = constructed from baronies inside the same province.
Duchies = province, keep province capital as duchy capital.
Kingdoms = States, unless the state is two small and user has disabled "small Kingdoms"

This PR addresses [Issue #7](https://github.com/pryvyd9/AzgaarToCK3/issues/7).

# Baronies
Baronies implement IProvince and are generated from burgs. Cells inside a province are assigned to its closest burg and this collection is a barony.

# Provinces
**Province** in Azgaar is a division of the map that is meant to represent subdivisions of states. 
But every cell is always assigned to a province.
Province 0 is a special province that represents a cell that does not lie within one of the other provinces. This can be because the cell is in the ocean, or because the state it lies in does not have a province covering the cell.

In Ck3 a province is something different. A province is a discreet region or area of the map. On land, this corresponds to a barony (or wasteland), in the ocean this is a sea tile and major river segment.

_Provinces have been reworked to be an interface to support this variability in what object can be a province._

# Calculating sea tiles and generating sea provinces
Sea province generation is being reworked to use the feature type to determine if it is a water tile or not. This has the advantage of also catching lakes (dry lakes, freshwater lakes, and sinkholes). This moves away from using the lack of statehood to determine if it is an ocean or not and avoids the issues with elevation that the Niefia generator experienced.
Sea tle generation will happen in much the same way as before where cells are merged to become larger regions of cells- The size of sea regions should be a parameter in settings.

# For developers and debug
Added a debug bool to the settings, this can be used to enable more detailed logging and to always answer yes to yes-no questions.

### CODE REVIEW

File: Converter/MapManager.cs [hunk 1]
```diff
@@ -1181,4 +1198,198 @@ public static async Task WriteHolySites(Map map, string[] pickedFaiths)
...
+    /// </summary>
+    /// <param name="burgs">Burgs to base the baronies on</param>
+    /// <returns>A list of baronies</returns>
+    private static List<Barony> GenerateBaronies(Dictionary<int, ConverterBurg> burgs)
+    {
+        // Next we instanciate a list of baronies. Since we know the final size of the list we can set the capacity to the number of burgs
+        List<Barony> baronies = new(burgs.Count - 1);
+        foreach (var burg in burgs.Skip(1)) //0'eth entry is always empty (See Azgaar data model)
+        {
+            baronies.Add(new Barony(burg.Value));
```
Comment (pryvyd9) [inline Converter/MapManager.cs]: when instantiating with final size you need to set values by index instead of adding them.

Comment (pryvyd9) [reply Converter/MapManager.cs]: you don't really need a foreach here.
`return burgs.Skip(1).Select(n => new Barony(n.Value)).ToList();`

Comment (MnTronslien) [reply Converter/MapManager.cs]: The list does not have a final size, I just initialize  it with an appropriate size for optimisation reasons. That way it does not have to resize so often.

Comment (MnTronslien) [reply Converter/MapManager.cs]: Might do a vanity pass when I am done, but I find having the foreach be explisit makes it easier for me to read and modify while working on it.

Comment (MnTronslien) [reply Converter/MapManager.cs]: Hey, thank you for taking the time to review btw!

**Comment by MnTronslien** [comment]
Todo: I need to make an ID manager for IProvinces. To ensure unique id's.
(or just remember to hand out Id's when all provinces are generated)

**Pull request closed without merging**